TLC5957.cpp

```diff
@@ -0,0 +1,366 @@
+{
+    uint64_t new_data = (td_bit_0 << 2) | (td_bit_1 << 3);
+    _function_data |= TD_SELECTION_MASK & new_data;
+}
+
+void TLC5957::setGroupDelaySelect(bool group_delay_bit)
+{
+    uint64_t new_data = group_delay_bit << 4;
+    _function_data |= GROUP_DELAY_SELECT_MASK & new_data;
+}
+
+void TLC5957::setRefreshMode(bool refresh_mode_bit)
+{
+    uint64_t new_data = refresh_mode_bit << 5;
+    _function_data |= REFRESH_MODE_MASK & new_data;
+}
+
+void TLC5957::setGclkEdgeSelect(bool gsclk_edge_bit)
+{
+    uint64_t new_data = gsclk_edge_bit << 6;
+    _function_data |= GSCLK_EDGE_SELECT_MASK & new_data;
+}
+
+void TLC5957::setPrechargeMode(bool precharge_bit)
+{
+    uint64_t new_data = precharge_bit << 7;
+    _function_data |= PRECHARGE_MODE_MASK & new_data;
+}
+
+void TLC5957::setEspwm(bool espwm_bit)
+{
+    uint64_t new_data = espwm_bit << 8;
+    _function_data |= ESPWM_MASK & new_data;
+}
+
+void TLC5957::setBlueCompensation(bool blue_compensation_bit)
+{
+    uint64_t new_data = blue_compensation_bit << 9;
+    _function_data |= BLUE_COMPENSATION_MASK & new_data;
+}
+
+void TLC5957::setSclkEdgeSelect(bool sclk_edge_bit)
+{
+    uint64_t new_data = sclk_edge_bit << 10;
+    _function_data |= SCLK_EDGE_SELECT & new_data;
+}
+
+void TLC5957::setLowGsEnhancement(bool low_gs_bit_0, bool low_gs_bit_1, bool low_gs_bit_3)
+{
+    uint64_t new_data = (low_gs_bit_0 << 11) | (low_gs_bit_1 << 12) | (low_gs_bit_3 << 13);
+    _function_data |= LOW_GS_ENHANCEMENT_MASK & new_data;
+}
+
+void TLC5957::setColorControl(uint16_t cc)
+{
+    setColorControl(cc, cc, cc);
+}
+
+void TLC5957::setColorControl(uint16_t ccr, uint16_t ccg, uint16_t ccb)
+{
+    uint64_t new_data = 0;
+
+    if (ccr > 511)
+        ccr = 511;
+    new_data |= ccr << 14;
+
+    if (ccg > 511)
+        ccg = 511;
+    new_data |= (uint32_t)ccg << 23;
+
+    if (ccb > 511)
+        ccb = 511;
+    new_data |= (uint64_t)ccb << 32;
+
+    _function_data |= COLOR_CONTROL_MASK & new_data;
+    _CC[0] = ccr;
+    _CC[1] = ccg;
+    _CC[2] = ccb;
+}
+
+void TLC5957::getColorControl(uint16_t* colorControl)
+{
+    colorControl[0] = _function_data >> 32 & 511;
+    colorControl[1] = _function_data >> 23 & 511;
+    colorControl[1] = _function_data >> 14 & 511;
+}
+
+void TLC5957::setBrightnessControl(uint8_t bc)
+{
+    if (bc > 7)
+        bc = 7;
+    uint64_t new_data = (uint64_t)bc << 41;
+    _function_data |= BRIGHTNESS_CONTROL_MASK & new_data;
+    _BC = bc;
+}
+
+uint8_t TLC5957::getBrightnessControl()
+{
+    return _function_data >> 41 & 7;
+}
+
+void TLC5957::setPokerMode(bool poker_mode_bit)
+{
+    uint64_t new_data = (uint64_t)poker_mode_bit << 44;
+    _function_data |= POKER_MODE_MASK & new_data;
+}
+
+void TLC5957::setFirstLineImprovement(bool first_line_bit_0,
+                                        bool first_line_bit_1,
+                                        bool first_line_bit_2)
+{
+    uint64_t new_data = ((uint64_t)first_line_bit_0 << 45)
```

Comment:
```suggestion
    uint64_t new_data = (((uint64_t)first_line_bit_0) << 45)
```

Make sure you first cast bools to uint64.

The issue is that `bool` is rather undefined. Many people hate the data type because when you use it for low level stuff it is:

1. Not guaranteed to be "small". The compiler can choose its size.
2. Nor is it guaranteed to be "efficient". Operating on "bools" is really undefined.


I'm not sure what part of the datasheet this refers to. But maybe you just want to store all 3 bits in an `uint8t`?

Reply:
if i'm going to cast to `uint64_t` anyways, should I just have that as the expected parameter type in the function signature?
instead of `uint8_t` since that is "too large" for the number of bits required as well

Reply:
Sure.